pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	s3UserHealthCheckName      = "rook-ceph-internal-s3-user-checker"
+	s3HealthCheckBucketName    = "rook-ceph-internal-s3-bucket-checker"
+	defaultStatusCheckInterval = 1 * time.Minute
+	s3HealthCheckObjectBody    = "Test Rook Object Data"
+	s3HealthCheckObjectBodyMD5 = "5f286306a2227a156ba770800e71b796"
+	s3HealthCheckObjectKey     = "rookObj1"
+	contentType                = "plain/text"
+)
+
+// bucketChecker aggregates the mon/cluster info needed to check the health of the monitors
+type bucketChecker struct {
+	context         *clusterd.Context
+	objContext      *Context
+	interval        time.Duration
+	serviceIP       string
+	port            string
+	client          client.Client
+	namespacedName  types.NamespacedName
+	healthCheckSpec *cephv1.HealthCheckSpec
+}
+
+// newbucketChecker creates a new HealthChecker object
+func newBucketChecker(context *clusterd.Context, objContext *Context, serviceIP, port string, client client.Client, namespacedName types.NamespacedName, healthCheckSpec *cephv1.HealthCheckSpec) *bucketChecker {
+	c := &bucketChecker{
+		context:         context,
+		objContext:      objContext,
+		interval:        defaultStatusCheckInterval,
+		serviceIP:       serviceIP,
```

Comment:
What if the external IPs are updated on the object CR? This service IP won't change ever, right? Since it will be redirected through the service IP to those IPs.

Reply:
Correct, the `serviceIP` is immutable only the endpoint IPs can change.